src/Microsoft.OData.Client/Serialization/HttpWebRequestMessage.cs

```diff
@@ -225,12 +238,74 @@ public override string GetHeader(string headerName)
         /// <param name="headerValue">Value of the header.</param>
         public override void SetHeader(string headerName, string headerValue)
         {
-#if DEBUG
-            // Only content length header is set after firing SendingRequest2 event
-            Debug.Assert(!this.sendingRequest2Fired || headerName == XmlConstants.HttpContentLength, "!this.sendingRequest2Fired || headerName == XmlConstants.HttpContentLength");
-#endif
-            Util.CheckArgumentNullAndEmpty(headerName, "headerName");
-            HttpWebRequestMessage.SetHeaderValue(this.httpRequest, headerName, headerValue);
+            if (_contentHeaderNames.Contains(headerName, StringComparer.OrdinalIgnoreCase))
+            {
+                // If this is a "Content" header then we cache the value (due
+                // to the message content not being set yet) and set it
+                // upon sending the message (when the content will be available)
+                if (string.Equals(headerName, XmlConstants.HttpContentType, StringComparison.OrdinalIgnoreCase))
+                {
+                    _contentHeaderValueCache["Content-Type"] = headerValue;
+                }
+                else if (string.Equals(headerName, XmlConstants.HttpContentLength, StringComparison.OrdinalIgnoreCase))
+                {
+                    SetHttpWebRequestContentLength(_contentHeaderValueCache, long.Parse(headerValue, CultureInfo.InvariantCulture));
+                }
+                else
+                {
+                    _contentHeaderValueCache[headerName] = headerValue;
+                }
+                return;
+            }
+
+            if (string.Equals(headerName, XmlConstants.HttpRequestAccept, StringComparison.OrdinalIgnoreCase))
+            {
+                _requestMessage.Headers.Remove(headerName);
+                _requestMessage.Headers.Add("Accept", headerValue);
+            }
+            else if (string.Equals(headerName, XmlConstants.HttpUserAgent, StringComparison.OrdinalIgnoreCase))
+            {
+                SetUserAgentHeader(_requestMessage, headerValue);
+            }
+            else if (string.Equals(headerName, XmlConstants.HttpAcceptCharset, StringComparison.OrdinalIgnoreCase))
+            {
+                SetAcceptCharset(_requestMessage, headerValue);
+            }
+            else
+            {
+                _requestMessage.Headers.Remove(headerName);
+                _requestMessage.Headers.Add(headerName, headerValue);
+            }
+        }
+
+        /// <summary>
+        /// Attempts to set the UserAgent property if it exists other wise returns false

```

Comment:
```suggestion
        /// Sets the user-agent value of the Http request message

```
The method only ever returns `void`

Reply:
RESOLVED